provider/server/exec.go

```diff
@@ -5,11 +5,150 @@ package server
+	return &iri.ExecResponse{
+		Url: s.buildURL("exec", token),
+	}, nil
+}
+
+func (s *Server) ServeExec(w http.ResponseWriter, req *http.Request, token string) {
+	ctx := req.Context()
+	log := logr.FromContextOrDiscard(ctx)
+
+	request, ok := s.execRequestCache.Consume(token)
+	if !ok {
+		log.V(1).Info("Rejecting unknown / expired token")
+		http.NotFound(w, req)
+		return
+	}
+
+	exec := executorExec{
+		Libvirt:         s.libvirt,
+		ExecRequest:     request,
+		VirshExecutable: s.virshExecutable,
+	}
+
+	handler, err := remotecommandserver.NewExecHandler(exec, remotecommandserver.ExecHandlerOptions{
+		StreamCreationTimeout: StreamCreationTimeout,
+		StreamIdleTimeout:     StreamIdleTimeout,
+	})
+	if err != nil {
+		log.Error(err, "Error creating exec handler")
+		code := http.StatusInternalServerError
+		http.Error(w, http.StatusText(code), code)
+		return
+	}
+
+	handler.Handle(w, req, remotecommandserver.ExecOptions{})
+}
+
+func (e executorExec) Exec(ctx context.Context, in io.Reader, out io.WriteCloser, resize remotecommand.TerminalSizeQueue) error {
+	var wg sync.WaitGroup
+
+	machineID := e.ExecRequest.MachineId
+	log := logr.FromContextOrDiscard(ctx).WithName(machineID)
+
+	_, err := e.Libvirt.DomainLookupByUUID(libvirtutils.UUIDStringToBytes(machineID))
+	if err != nil {
+		if !libvirtutils.IsErrorCode(err, libvirt.ErrNoDomain) {
+			return fmt.Errorf("error looking up domain: %w", err)
+		}
+		return fmt.Errorf("machine %s has not yet been synced", machineID)
+	}
+
+	uri, err := e.Libvirt.ConnectGetUri()
+	if err != nil {
+		return fmt.Errorf("error getting connection uri")
+	}
+
+	cmd := exec.CommandContext(ctx, e.VirshExecutable, "-c", strings.TrimSpace(uri), "console", machineID)
+
+	f, err := pty.Start(cmd)
+	if err != nil {
+		return fmt.Errorf("error starting command: %w", err)
+	}
+
+	// Wrap the input stream with an escape proxy. Escape Sequence Ctrl + ] = 29
+	inputReader := term.NewEscapeProxy(in, []byte{29})
+
+	wg.Add(2)
+	// ReadInput: go routine to read the input from the reader, and write to the terminal.
+	go func() {
+		defer wg.Done()
+
+		buf := make([]byte, 1024)
+		for {
+			n, err := inputReader.Read(buf)
+			if err != nil {
+				if _, ok := err.(term.EscapeError); ok {
+					_, _ = f.Write(buf) // This is to close the writer.
+					log.Info("Closed reading the terminal. Escape sequence received")
+					return
+				}
+				log.Error(err, "error reading bytes")
+				return
+			}
+
+			_, err = f.Write(buf[:n])
+			if err != nil {
+				log.Error(err, "error writing to the file descriptor")
+				return
+			}
+		}
+	}()
+
+	// WriteOutput: go routine for writing the output back to the Writer.
+	go func() {
+		defer wg.Done()
+		_, _ = io.Copy(out, f)
+		log.Info("Closed writing to the terminal")
+	}()
+
+	if err := cmd.Wait(); err != nil {
+		// Avoid returning so that the function can verify if all go routines are terminated.
+		log.Error(err, "Console command interrupted")
```

Comment:
you don't have same format of error message. You combine uppercase and lowercase first letter for messages. I suggest use one format ideally lowercase: Primary error has to be lowercases which are part of log output, so i recommend follow same format: https://google.github.io/styleguide/go/decisions.html#error-strings